### PR TITLE
DA-283 Rollback to specific version

### PR DESCRIPTION
**Pull request by momoneko**
## Ready State
**Ready**
## Synopsis
Often times it is required to rollback a service to a specific AMI, like a few versions back of some services. 
This PR adds a `--rollback-to` option to `ef-version` to enable re-setting the current version to a version in the history of the service.
[DA-283](https://ellation.atlassian.net/browse/DA-283)
## Changelog
- add a `--rollback-to` flag
- add rollback to an AMI functionality to cmd_rollback
- test cover cmd_rollback
## Testing
1. Run the unit-tests
For a real-life test, run
```
ef-version --history text playheads ami-id alpha0
```
pick a previous version and use the current code-base to run
```
ef_version playheads ami-id alpha0 --rollback-to <ami-version> --devel --commit
```

### CODE REVIEW

File: efopen/ef_version.py [hunk 1]
```diff
@@ -554,21 +556,53 @@ def cmd_history(context):
...
   """
-  last_stable = get_versions(context, return_stable=True)
-  if len(last_stable) != 1:
-    fail("Didn't find a version marked stable for key: {} in env/service: {}/{}".format(
-         context.key, context.env, context.service_name))
-  context.value = last_stable[0].value
-  context.commit_hash = last_stable[0].commit_hash
-  context.build_number = last_stable[0].build_number
-  context.location = last_stable[0].location
+  if isinstance(context.rollback, bool):
```
Comment (triplepoint) [inline efopen/ef_version.py]: Is this how we want to do this?  Testing on types can get confusing, and can be brittle in dynamic languages like Python.

Why not have the `rollback` and `rollback_to` arguments remain separate variables, in a mutually exclusive ArgParse group.  Then the logic is to test on whether rollback_to has a value, and if not test whether rollback is True.

File: efopen/ef_version.py [hunk 2]
```diff
@@ -322,7 +324,7 @@ def handle_args_and_set_context(args):
     fail("Error in --limit. Valid range: 1..1000")
   context._limit = parsed_args["limit"]
   context._location = parsed_args["location"]
-  context._rollback = parsed_args["rollback"]
+  context._rollback = parsed_args["rollback"] or parsed_args["rollback_to"]
```
[the remark this answers is not in the document]
Comment (momoneko) [reply efopen/ef_version.py]: These options already are in a mutually exclusive group.
https://github.com/crunchyroll/ef-open/blob/5be662bbfb5e257d331155b5eb4d514a0d33946f/efopen/ef_version.py#L274-L280

File: efopen/ef_version.py (hunk 1 above)
Comment (vfangellation) [reply efopen/ef_version.py]: Yea I think this makes more sense to keep them as two separate variables

File: tests/unit_tests/test_ef_version.py [hunk 3]
```diff
@@ -200,3 +218,153 @@ def test_precheck_dist_hash_version_none(self, mock_version_object):
...
     mock_version_object.side_effect = ClientError(response, "Get Object")
     self.assertTrue(ef_version.precheck_dist_hash(self))
+
+class TestEFVersionModule(unittest.TestCase):
+
+  def setUp(self):
+    self.versions = map(
+        ef_version.Version, [
+            {
+                u'Body': StringIO.StringIO("ami-0f85b8e7ca0788951"),
```
Comment (vfangellation) [inline tests/unit_tests/test_ef_version.py]: what's with all the u's?

File: efopen/ef_version.py [hunk 4]
```diff
@@ -554,6 +562,37 @@ def cmd_history(context):
     print(json.dumps(versions, cls=VersionEncoder))
 
 
+def get_latest_stable_version(context):
+    """
+    Get the latest stable version from the version registry
+    Args:
+      context: a populated EFVersionContext object
+    """
+    last_stable = get_versions(context, return_stable=True)
```
Comment (vfangellation) [inline efopen/ef_version.py]: I kinda don't like the name of this variable. The return value are all the stable versions given the context and we return the version[0] as the most recent stable one.

You could store the result in versions and then have
last_stable = versions[0] and return that var
I find it perfectly reasonable that you create temp vars to help the clarity of the code.

Comment (momoneko) [reply efopen/ef_version.py]: This was copy/pasted from `cmd_rollback` and is no longer used. Will remove this function.

**Pull request merged**